src/vision/status_quo/alan_iteratively_regresses.md

```diff
@@ -0,0 +1,113 @@
+
+For the other two types, Alan responds that the `SuccessCode` is small, and that the error variants are never constructed in his benchmark code. Barbara explains that the size of `Result<T, E>` has to be large enough to hold either variant, and that `memcpy`'ing a result is going to move all of those bytes. Alan investigates and sees that `DistriErr` has variants that embed byte arrays that go up to 50kb in size. Barbara recommends that Alan look into boxing the variants, or the whole `DistriErr` type itself, in order to reduce the cost of moving it around.
+
+Alan uses Barbara's feedback to box some of the data, and this cuts the `memcpy` traffic in the `perf report` to one quarter of what it had been reporting previously.
+
+However, there remains a significant performance delta between the Java version and the Rust version. Alan is not sure his Rust-rewrite attempt is going to get anywhere beyond the prototype stage.
+
+## 🤔 Frequently Asked Questions
+
+*Here are some standard FAQ to get you started. Feel free to add more!*
+
+### **What are the morals of the story?**
+
+1. Rust promises great performance, but when performance is not meeting one's targets, it is hard to know what to do next. Rust mostly leans on leveraging existing tools for native code development, but those tools are (a.) foreign to many of our developers, (b.) do not always measure up to what our developers have access to elsewhere, (c.) do not integrate as well with Rust as they might with C or C++.
+
+2. Lack of certain language features leads developers to use constructs like `#[async_trait]` which add performance overhead that is (a.) hard to understand and (b.) may be significant.
+
+3. Rust makes some things very explicit, e.g. the distinction between `Box<T>` versus `T` is quite prominent. But Rust's expressive type system also makes it easy to compose types without realizing how large they have gotten.
+
+4. Programmers do not always have a good mental model for where expensive moves are coming from.
+
+5. An important specific instance of (1c.) for the async vision: Native code tools do not have any insight into Rust's async model, as that is even more distant from the execution model of C and C++.
+
+6. We can actually generalize (5.) further: When async performance does not match expectations, developers do not have much insight into whether the performance pitfalls arise from issues deep in the async executor that they have selected, or if the problems come directly from overheads built into the code they themselves have written.
+
+### **What are the sources for this story?**
+
+Discussions with engineers at Amazon Web Services.
+
+### **Why did you choose Alan to tell this story?**
+
+I chose Alan because he is used to Java, where these issues play out differently.
+
+Java has very mature tooling, including for performance investigations. Alan is frustrated by his attempts to use (or even identify) equivalent tools for Rust.
```

Comment:
Is it worth mentioning what java tools Alan would have reached for in this case?  (Would he have used JConsole/VisualVM/JProfiler?  Eclipse MAT?  Something even more advanced?)

Reply:
I myself am not an expert in modern Java Profiler technologies. I suppose the main question is whether Alan would reach for a commerical tool like  [JProfiler](https://www.ej-technologies.com/products/jprofiler/overview.html), or if they would only use freely available tools like [VisualVM](https://visualvm.github.io/). Maybe I'll just link to both.

Reply:
addressed in bcce67bcda80282f1bd96744055d4b87176a88f4